# Homework 3
Simple Word Autocomplete / Autocorrect and Word Probability

1. Choose a English corpus from NLTK. Identify all the unique words in your corpus (types). This is your program's vocabulary. Compute the length of your corpus (N): total number of words in the corpus (tokens).

We first import the corpus of our choosing. A relatively short corpus is previous president Barak Obama's 2013 inaugural speech.

In [107]:
import nltk
from nltk.corpus import inaugural

nltk.download('inaugural')
obama_speech = inaugural.words(fileids='2013-Obama.txt')

[nltk_data] Downloading package inaugural to
[nltk_data]     /Users/leavism/nltk_data...
[nltk_data]   Package inaugural is already up-to-date!


The corpus reader provided with NLTK has a variety of methods to read data from a corpus. In our case above, we are obtaining a list of words, but below are some examples that can read the corpus as raw text, a list of sentences, or a list of paragraphs.

In [108]:
%%script echo Skipping
inaugural.raw('1789-Washington.txt')
# 'Fellow-Citizens of the Senate ...'
inaugural.sents('1789-Washington.txt')
# [['Fellow', '-', 'Citizens'...], ['Among', 'the', 'vicissitudes'...]...]
inaugural.paras('1789-Washington.txt')
# [[['Fellow', '-', 'Citizens'...]],
# [['Among', 'the', 'vicissitudes'...],
# ['On', 'the', 'one', 'hand', ',', 'I'...]...]...]

Skipping


To determine the number of unique words in the corpus (types), we can use the `set()` function to convert a Python iterable, like a list, into a set data type. A set is a collection of unique objects that is unordered, therefore a set of the inaugural speech will only contain unique types of the corpus.

We can also count the length of the corpus by calling Python's `len()` function on the provided list.

In [109]:
types = set(obama_speech)

tokens = len(obama_speech)

print("President Barrak Obama's 2013 Inaugural Speech")
print("Types:", len(types))
print("Tokens:", tokens)

President Barrak Obama's 2013 Inaugural Speech
Types: 820
Tokens: 2369


2. Find the number of occurrences (frequency) of each type.

There are two ways to approach this problem. We can use the `Counter` class, a subclass of the dictionary, which will find the number of occurrences of each type. But I'm sure Dr. Kulkarni would want us to try it without using `Counter` too. I'll provide both, but my notebook will keep the results of the `Counter` class.

In [110]:
from collections import Counter

word_freq = Counter(obama_speech)

for word, freq in word_freq.items():
    print(f"{word}: {freq}")

Thank: 3
you: 5
.: 91
so: 6
much: 1
Vice: 1
President: 2
Biden: 1
,: 99
Mr: 1
Chief: 1
Justice: 1
Members: 1
of: 69
the: 100
United: 2
States: 2
Congress: 1
distinguished: 1
guests: 1
and: 85
fellow: 3
citizens: 6
:: 5
Each: 1
time: 9
we: 46
gather: 1
to: 66
inaugurate: 1
a: 36
bear: 1
witness: 1
enduring: 3
strength: 2
our: 69
Constitution: 1
We: 23
affirm: 1
promise: 2
democracy: 2
recall: 1
that: 51
what: 7
binds: 1
this: 12
Nation: 1
together: 4
is: 25
not: 23
colors: 1
skin: 1
or: 13
tenets: 1
faith: 1
origins: 1
names: 1
What: 1
makes: 2
us: 21
exceptionalâ: 1
€”: 13
Americanâ: 1
allegiance: 1
an: 6
idea: 1
articulated: 1
in: 21
declaration: 1
made: 3
more: 7
than: 6
two: 1
centuries: 2
ago: 1
hold: 1
these: 10
truths: 2
be: 10
self: 3
-: 9
evident: 3
all: 12
men: 3
are: 21
created: 3
equal: 5
;: 14
they: 7
endowed: 1
by: 11
their: 4
Creator: 1
with: 12
certain: 1
unalienable: 1
rights: 2
among: 1
life: 4
liberty: 5
pursuit: 2
happiness: 4
Today: 1
continue: 1
never: 3
ending: 2


In [111]:
%%script echo Skipping
word_freq = {}
for word in obama_speech:
    if word in types:
        if word in word_freq:
            word_freq[word] += 1
        else:
            word_freq[word] = 1

# Print the frequency of each word
for word, freq in word_freq.items():
    print(f"{word}: {freq}")

Skipping


3. Find the relative frequency of each type, $ W $, where relative frequency of `W = frequency_of_W / N`. This relative frequency is the probability (likelihood) of each type in the corpus.

For every word in the corpus, divide the frequency of the word by the total number of tokens. For example for the word "Thank" that occurs 3 times, the relative frequency is calculated by $ w = 3 / 2369 = 0.0012663571 $. Keep in mind that tokens like `.`, `:`, and the coma `,` are counted as unique types.

In [112]:
relative_freq = {}
for word, freq in word_freq.items():
    relative_freq[word] = freq / tokens

# Print the relative frequency of each word
for word, rel_freq in relative_freq.items():
    print(f"{word}: {rel_freq}")

Thank: 0.001266357112705783
you: 0.002110595187842972
.: 0.038412832418742084
so: 0.002532714225411566
much: 0.00042211903756859433
Vice: 0.00042211903756859433
President: 0.0008442380751371887
Biden: 0.00042211903756859433
,: 0.04178978471929084
Mr: 0.00042211903756859433
Chief: 0.00042211903756859433
Justice: 0.00042211903756859433
Members: 0.00042211903756859433
of: 0.02912621359223301
the: 0.04221190375685943
United: 0.0008442380751371887
States: 0.0008442380751371887
Congress: 0.00042211903756859433
distinguished: 0.00042211903756859433
guests: 0.00042211903756859433
and: 0.03588011819333052
fellow: 0.001266357112705783
citizens: 0.002532714225411566
:: 0.002110595187842972
Each: 0.00042211903756859433
time: 0.003799071338117349
we: 0.019417475728155338
gather: 0.00042211903756859433
to: 0.027859856479527226
inaugurate: 0.00042211903756859433
a: 0.015196285352469396
bear: 0.00042211903756859433
witness: 0.00042211903756859433
enduring: 0.001266357112705783
strength: 0.000844238075

4. For every input string XYZ:

    4.a. If the input string XYZ exists in your vocabulary, return "XYZ is a complete and correct word as per corpus ___, and its probability is __"

    4.b. If the input string doesn't exist in your vocabulary, perform the below steps
    
    4.b.i. Calculate the similarity between the input string and each type using Levenshtein distance.

    4.b.ii. Output 5 types that are most similar to the input string as per the Levenshtein distance. Also, output the probability for each of the 5 types.

Firstly, lets write a function that takes the user input XYZ and calculates its similarity to every type according to the Levenshtein distance. First we'll import the Levenshtein Python C extension module.

In [113]:
import Levenshtein

The Levenshtein module makes this task really. I call the `distance()` function on every type in the corpus and insert the result as a tuple of the word and edit distance into a list. I then sort the resulting list and output the top 5 (this value can be changed at the time of calling this function) similiars types to the input string.

In [114]:
def get_most_similar_words(input_word, types, n=5):
    similar_words = []
    for word in types:
        distance = Levenshtein.distance(input_word, word)
        similar_words.append((word, distance))
    similar_words.sort(key=lambda x: x[1])
    return similar_words[:n]

I then take the user input, sanitize it by removing leading or trailing whitespace, and then check whether the user input is already in the collection of types. If it exists, I print that it exists in the corpus and its relative probability. If it does not exist, I print the five most similar types and the edit distance.

In [115]:
user_input = input("Enter a word: ").strip()

if user_input in types:
    probability = relative_freq[user_input]
    print(f"{user_input} is a complete and correct word as per corpus \"2013-obama.txt\", and its probability is {probability}")
else:
    similar_words = get_most_similar_words(user_input, types)
    print(f"User input: {user_input}")
    for word, distance in similar_words:
        probability = relative_freq[word]
        print(f"{word} (Similarity: {distance}, Probability: {probability})")

Thank is a complete and correct word as per corpus "2013-obama.txt", and its probability is 0.001266357112705783


# Summary

This assignment was fun! And less tedious than the previous homeworks, but I do understand the importance of reading code to learn code. I've written my thoughts and explained my code throughout this notebook, cell by cell. Although the meat of the logic is provided by the Levenshtein module that was imported, one place of improvement is clarifying the definition of word. This iteration of my code considers `.`, `:`, `,` and other forms of punctuation as types. Morever, it considers the parts of contractions as types, like `ll` and `s`. Removing these as unique types would give a more accurate relative frequency in a more traditional sense of a "word."